text/0625-helper-managers.md

```diff
@@ -0,0 +1,504 @@
+- called immediately after the `createHelper` hook is called
+
+**Never**
+- called if the `hasDestroyable` capability is disabled
+
+### Capabilities
+
+There are three proposed capabilities for helper managers:
+
+* `hasDestroyable`
+* `isScheduledEffect`
+* `hasValue`
+
+Out of these capabilities, one of `isScheduledEffect` or `hasValue` _must_ be
+enabled. The other must _not_ be enabled, meaning they are mutually exclusive.
+
+#### `hasDestroyable`
+
+Determines if the helper has a destroyable to include in the destructor
+hierarchy. If enabled, the `getDestroyable` hook will be called, and its result
+will be associated with the destroyable parent block.
+
+#### `hasValue`
+
+Determines if the helper has a value which can be used externally. The helper's
+`getValue` hook will be run whenever the value of the helper is accessed if this
+capability is enabled.
+
+#### `isScheduledEffect`
+
+Determines if the helper is scheduled. If enabled, the helper's `runEffect` hook
+will run after render, and will not allow any type of state mutation when
+running.
+
+### Scheduled Helpers Timing
+
+Scheduled helpers run their effects after render, and after modifiers have been
+applied for a given render, but before paint. The exact timing may shift around,
+and may or may not correspond to a single rendering pass in cases where there
+are multiple rendering passes in a single paint.
+
+In the future different timings may be added as options for scheduling. For
+instance, a timing to call the effect using `requestIdleCallback`, when the
+browser has finished rendering and handling higher priority work, could be
+added. However, this is out of scope for this RFC.
+
+## How we teach this
+
+Helper managers are a low-level construct that is generally only meant to be
+used by experts and addon authors. As such, it will only be taught through API
+documentation. In addition, for precision and clarity, the API docs will include
+snippets of TypeScript interfaces where appropriate.
+
+### API Docs
+
+#### `setHelperManager`
+
+Sets the helper manager for an object or function.
+
+```js
+setHelperManager((owner) => new ClassHelperManager(owner), Helper.prototype)
+```
+
+When a value is used as a helper in a template, the helper manager is looked up
+by on the object by walking up its prototype chain and finding the first helper
+manager. This manager then receives the value and can create and manage an
+instance of a helper from it. This provides a layer of indirection that allows
+users to design high-level helper APIs, without Ember needing to worry about the
+details. High-level APIs can be experimented with and iterated on while the
+core of Ember helpers remains stable, and new APIs can be introduced gradually
+over time to existing code bases.
+
+`setHelperManager` receives two arguments:
+
+1. A factory function, which receives the `owner` and returns an instance of a
+   helper manager.
+2. The object to associate the factory function with.
```

Comment:
Would you mind making it clear that `function` is also allowed? I know the TS interface makes it clear but it is easy to miss, and doesn't seem to hurt to say "function or object" here.

Reply:
I think you should also tweak this to say "helper definition" to make it a easier association with the TS interfaces you drew up above.

```suggestion
2. A helper definition, which is the object or function to associate the factory function with.
```